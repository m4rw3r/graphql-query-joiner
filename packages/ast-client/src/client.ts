import type { GraphQLResponse, GraphQLError, Query, RenameMap } from "./query";
import type { QueryBundle } from "./bundle";

import { print } from "graphql/language";
import { createBundle, createDocument, mergeBundle } from "./bundle";
import {
  missingVariableError,
  parseError,
  queryError,
  requestError,
} from "./error";

export type QueryRunner = (query: {
  query: string;
  variables: Record<string, unknown>;
}) => Promise<GraphQLResponse<unknown>>;

export type ClientArgs = {
  runQuery: QueryRunner;
  debounce?: number;
};

export type Client<O> = <P, R>(
  query: Query<P, R>,
  variables: P,
  options?: Readonly<O>,
) => Promise<R>;

type ResolveFn<T> = (value: T) => void;
type RejectFn = (error: unknown) => void;

export type Group = {
  bundle: QueryBundle;
  variables: Record<string, unknown>;
  queries: Array<GroupedQuery>;
};

export type GroupedQuery = {
  renamedFields: RenameMap;
  resolve: ResolveFn<unknown>;
  reject: RejectFn;
};

export type GroupedQueryWithError = GroupedQuery & {
  errors: Array<GraphQLError>;
};

const resolved: Promise<void> = new Promise((resolve) => resolve(undefined));

function setVariable(
  variables: Record<string, unknown>,
  parameters: unknown,
  nameInAst: string,
  nameInQuery: string,
): void {
  if (
    typeof parameters === "object" &&
    parameters &&
    Object.prototype.hasOwnProperty.call(parameters, nameInAst)
  ) {
    // SAFETY: If we are requesting variables and pass these checks, we can
    // assume it is a correct variable typed based on the Query
    variables[nameInQuery] = (parameters as Record<string, unknown>)[nameInAst];
  } else {
    throw missingVariableError(nameInAst);
  }
}

function createGroup<P, R>(
  bundle: QueryBundle,
  parameters: P,
  resolve: ResolveFn<R>,
  reject: RejectFn,
): Group {
  const variables = {};
  const firstBundleFields: Record<string, string> = {};

  for (const [k] of bundle.variables) {
    setVariable(variables, parameters, k, k);
  }

  for (const [k] of bundle.fields) {
    firstBundleFields[k] = k;
  }

  return {
    bundle,
    variables,
    queries: [
      {
        renamedFields: firstBundleFields,
        // SAFETY: We will only call this in runGroup() with data extracted
        // from the fields above
        resolve: resolve as ResolveFn<unknown>,
        reject,
      },
    ],
  };
}

/**
 * Handles a fetch-Response and parses it into a GraphQLResponse,
 * throws if the request is not ok or if JSON fails to parse.
 */
export async function handleFetchResponse<R>(response: Response): Promise<R> {
  const bodyText = await response.text();

  if (!response.ok) {
    throw requestError(
      response,
      bodyText,
      `Received status code ${response.status}`,
    );
  }

  try {
    // SAFETY: Since it is successful we assume we have GraphQL-data in the
    // correct format
    return JSON.parse(bodyText) as R;
  } catch (error) {
    throw parseError(response, bodyText, error);
  }
}

export function enqueue<P, R>(
  pending: Array<Group>,
  newBundle: QueryBundle,
  parameters: P,
  resolve: ResolveFn<R>,
  reject: RejectFn,
): void {
  const last = pending[pending.length - 1];

  if (last && last.bundle.operation === newBundle.operation) {
    // Merge the queries since their types match
    const { bundle, renamedVariables, renamedFields } = mergeBundle(
      last.bundle,
      newBundle,
    );
    last.bundle = bundle;

    for (const [k, v] of Object.entries(renamedVariables)) {
      setVariable(last.variables, parameters, k, v);
    }

    last.queries.push({
      renamedFields,
      // SAFETY: We will only call this in runGroup() with data extracted
      // from the fields above
      resolve: resolve as ResolveFn<unknown>,
      reject,
    });
  } else {
    pending.push(createGroup(newBundle, parameters, resolve, reject));
  }
}

/**
 * Matches errors into groups based on their path.
 *
 * If an error is not in any of the supplied groups then it is assigned to all
 * of them.
 */
export function groupErrors<T extends { renamedFields: RenameMap }>(
  errors: Array<GraphQLError>,
  groups: Array<T>,
): Array<T & { errors: Array<GraphQLError> }> {
  const groupsWithErrors = groups.map((group) => ({
    ...group,
    // TODO: Maybe use something more efficent
    errors: errors.filter(
      (error) =>
        Array.isArray(error.path) &&
        error.path.length > 0 &&
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        Object.values(group.renamedFields).includes(error.path[0]!),
    ),
  }));

  // We need to make sure that errors without a path also gets propagated
  const missingErrors = errors.filter(
    (error) => !groupsWithErrors.some((g) => g.errors.includes(error)),
  );

  groupsWithErrors.forEach((group) => group.errors.push(...missingErrors));

  return groupsWithErrors;
}

export async function runGroup(
  runQuery: QueryRunner,
  { bundle, variables, queries }: Group,
): Promise<void> {
  try {
    const bundledResponse = await runQuery({
      query: print(createDocument(bundle)),
      variables,
    });

    for (const { renamedFields, resolve, reject, errors } of groupErrors(
      bundledResponse.errors || [],
      queries,
    )) {
      const data: Record<string, unknown> = {};

      if ("data" in bundledResponse) {
        for (const [k, v] of Object.entries(renamedFields)) {
          // SAFETY: Since we have a data-field we assume that it contains
          // the requested keys
          data[k] = (bundledResponse.data as Record<string, unknown>)[v];
        }
      }

      if (errors.length > 0) {
        reject(queryError(errors, data));
      } else {
        resolve(data);
      }
    }
  } catch (e) {
    queries.forEach(({ reject }) => reject(e));
  }
}

export function runGroups(
  runQuery: QueryRunner,
  groups: Array<Group>,
): Promise<void> {
  return groups.reduce(
    (p, group) => p.then(() => runGroup(runQuery, group)),
    resolved,
  );
}

// TODO: Rewrite as class
export function createClient({
  runQuery,
  debounce = 50,
}: ClientArgs): Client<object> {
  let next = resolved;
  let timer: ReturnType<typeof setTimeout> | undefined;
  let pending: Array<Group> = [];

  const fire = (): void => {
    // TODO: Control how this chaining is done?
    next = next.then((): Promise<void> => {
      const r = runGroups(runQuery, pending);

      // Clear now that we fired off stuff
      pending = [];
      timer = undefined;

      return r;
    });
  };

  return <P, R>(query: Query<P, R>, parameters: P): Promise<R> =>
    new Promise<R>((resolve: ResolveFn<R>, reject: RejectFn): void => {
      enqueue(pending, createBundle(query), parameters, resolve, reject);

      if (!timer) {
        timer = setTimeout(fire, debounce);
      }
    });
}
