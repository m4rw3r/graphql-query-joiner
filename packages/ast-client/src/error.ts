import type { GraphQLError } from "./query";

export interface MissingVariableError extends Error {
  name: "MissingVariableError";
  variableName: string;
}

export interface RequestError extends Error {
  name: "RequestError";
  response: Response;
  statusCode: number;
  bodyText: string;
}

export interface ParseError extends Error {
  name: "ParseError";
  response: Response;
  statusCode: number;
  bodyText: string;
}

export interface QueryError extends Error {
  name: "QueryError";
  errors: GraphQLError[];
  queryData: Record<string, unknown>;
}

function graphqlErrorMessage({ message }: GraphQLError): string {
  return message;
}

export function requestError(
  response: Response,
  bodyText: string,
  message: unknown,
): RequestError {
  const error: RequestError = new Error(message as string) as RequestError;

  error.name = "RequestError";
  error.response = response;
  error.statusCode = response.status;
  error.bodyText = bodyText;

  return error;
}

export function parseError(
  response: Response,
  bodyText: string,
  message: unknown,
): ParseError {
  const error: ParseError = new Error(message as string) as ParseError;

  error.name = "ParseError";
  error.response = response;
  error.statusCode = response.status;
  error.bodyText = bodyText;

  return error;
}

export function queryError(
  errors: GraphQLError[],
  queryData: Record<string, unknown>,
): QueryError {
  const error: QueryError = new Error(
    errors.map(graphqlErrorMessage).join(", "),
  ) as QueryError;

  error.name = "QueryError";
  error.errors = errors;
  error.queryData = queryData;

  return error;
}

export function missingVariableError(name: string): MissingVariableError {
  const error: MissingVariableError = new Error(
    `Variable '${name}' is missing.`,
  ) as MissingVariableError;

  error.name = "MissingVariableError";
  error.variableName = name;

  return error;
}
