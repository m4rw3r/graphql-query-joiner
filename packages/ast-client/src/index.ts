export type {
  Query,
  QueryParameters,
  QueryResult,
  GraphQLError,
  GraphQLResponse,
} from "./query";
export type {
  Client,
  CreateClientOptions,
  PreparedQuery,
  QueryRunner,
} from "./client";

export { createClient, handleFetchResponse } from "./client";
