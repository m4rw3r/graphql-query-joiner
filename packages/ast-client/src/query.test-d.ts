// File testing types
//
// See https://tstyche.org/

import type {
  EmptyObject,
  Mutation,
  Operation,
  OperationParameters,
  OperationResult,
  Query,
} from "./query";
import type { Client } from "./client";
import type { DocumentNode } from "graphql/language";
import { describe, expect, test } from "tstyche";

interface MyQueryParams {
  path: string;
}

interface MyQueryResult {
  id: number;
  content: string;
}

// Dummy client
const runQuery: Client = <O extends Operation<any, any>>(
  _query: O,
  _params?: OperationParameters<O> | EmptyObject,
): Promise<OperationResult<O>> => ({}) as Promise<OperationResult<O>>;

const myQuery = {} as Query<MyQueryParams, MyQueryResult>;
const voidQuery = {} as Query<void, MyQueryResult>;
const undefinedQuery = {} as Query<undefined, MyQueryResult>;
const emptyQuery = {} as Query<EmptyObject, MyQueryResult>;
const emptyRecordQuery = {} as Query<Record<string, never>, MyQueryResult>;

// TODO: Query vs Mutation vs Operation

describe("Queries", () => {
  test("are not plain document nodes", () => {
    expect(myQuery).type.not.toEqual<DocumentNode>();
    expect(voidQuery).type.not.toEqual<DocumentNode>();
    expect(undefinedQuery).type.not.toEqual<DocumentNode>();
    expect(emptyQuery).type.not.toEqual<DocumentNode>();
    expect(emptyRecordQuery).type.not.toEqual<DocumentNode>();
  });

  test("All queries are DocumentNodes", () => {
    expect<DocumentNode>().type.toBeAssignable(myQuery);
    expect<DocumentNode>().type.toBeAssignable(voidQuery);
    expect<DocumentNode>().type.toBeAssignable(undefinedQuery);
    expect<DocumentNode>().type.toBeAssignable(emptyQuery);
    expect<DocumentNode>().type.toBeAssignable(emptyRecordQuery);
  });

  test("But DocumentNodes are not queries", () => {
    expect(myQuery).type.not.toBeAssignable<DocumentNode>();
    expect(voidQuery).type.not.toBeAssignable<DocumentNode>();
    expect(undefinedQuery).type.not.toBeAssignable<DocumentNode>();
    expect(emptyQuery).type.not.toBeAssignable<DocumentNode>();
    expect(emptyRecordQuery).type.not.toBeAssignable<DocumentNode>();
  });

  test("are not Mutations", () => {
    expect(myQuery).type.not.toBeAssignable<Mutation<any, any>>();
    expect(voidQuery).type.not.toBeAssignable<Mutation<any, any>>();
    expect(undefinedQuery).type.not.toBeAssignable<Mutation<any, any>>();
    expect(emptyQuery).type.not.toBeAssignable<Mutation<any, any>>();
    expect(emptyRecordQuery).type.not.toBeAssignable<Mutation<any, any>>();
  });
});

// prettier-ignore
describe("Client", () => {
  test("If the Query has a non-empty parameter object, the variables parameter is required", () => {
    expect(runQuery(myQuery, { path: "foo" })).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(myQuery)).type.toRaiseError(2554); // "Expected 2 arguments, but got 1."
  });

  test("If the query does not require any parameters, we can skip them", () => {
    expect(runQuery(voidQuery, undefined)).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(voidQuery)).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(undefinedQuery, undefined)).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(undefinedQuery)).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(emptyQuery)).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(emptyQuery, {})).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(emptyRecordQuery)).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(emptyRecordQuery, {})).type.toEqual<Promise<MyQueryResult>>();
  });

  test("Empty parameters also accept undefined or empty object", () => {
    expect(runQuery(voidQuery, undefined)).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(voidQuery, {})).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(undefinedQuery, undefined)).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(undefinedQuery, {})).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(emptyQuery, undefined)).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(emptyQuery, {})).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(emptyRecordQuery, undefined)).type.toEqual<Promise<MyQueryResult>>();
    expect(runQuery(emptyRecordQuery, {})).type.toEqual<Promise<MyQueryResult>>();
  });

  // TODO: Should it though? This could cause issues later when using
  test("Extra parameter should raise error on Query<void, _>", () => {
    expect(runQuery(voidQuery, { test: "foo" })).type.toRaiseError(2322); // "Type string is not assignable to never."
  });

  test("Extra parameter should raise error on Query<undefined, _>", () => {
    expect(runQuery(undefinedQuery, { test: "foo" })).type.toRaiseError(2322); // "Type string is not assignable to never."
  });

  test("Extra parameter should raise error on Query<EmptyObject, _>", () => {
    expect(runQuery(emptyQuery, { test: "foo" })).type.toRaiseError(2322); // "Type string is not assignable to never."
  });

  test("Extra parameter should raise error on Query<Record<string, never>, _>", () => {
    expect(runQuery(emptyRecordQuery, { test: "foo" })).type.toRaiseError(2322); // "Type string is not assignable to never."
  });
});

describe("OperationResult", () => {
  test("OperationResult should resolve to the type of the result", () => {
    expect<OperationResult<typeof myQuery>>().type.toEqual<MyQueryResult>;
    expect<OperationResult<typeof voidQuery>>().type.toEqual<MyQueryResult>;
    expect<OperationResult<typeof undefinedQuery>>().type
      .toEqual<MyQueryResult>;
    expect<OperationResult<typeof emptyQuery>>().type.toEqual<MyQueryResult>;
    expect<OperationResult<typeof emptyRecordQuery>>().type
      .toEqual<MyQueryResult>;
  });
});

describe("OperationParameters", () => {
  test("OperationParameters should resolve to the type of the parameters", () => {
    expect<OperationParameters<typeof myQuery>>().type.toEqual<MyQueryParams>;
    expect<OperationParameters<typeof voidQuery>>().type.toEqual<void>;
    expect<OperationParameters<typeof undefinedQuery>>().type
      .toEqual<undefined>;
    expect<OperationParameters<typeof emptyQuery>>().type.toEqual<EmptyObject>;
    expect<OperationParameters<typeof emptyRecordQuery>>().type
      .toEqual<EmptyObject>;
  });
});
