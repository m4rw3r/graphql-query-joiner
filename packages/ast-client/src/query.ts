//import type { DocumentNode } from "graphql/language";
import type {
  ResultOf,
  TypedDocumentNode,
  VariablesOf,
} from "@graphql-typed-document-node/core";

/**
 * Helper type to differentiate a Query from a Mutation.
 *
 * @deprecated
 */
export type Query<P, R> = Operation<P, R> & {
  /**
   * Phantom type (never assigned) for indicating it is a query.
   */
  readonly _O: "query";
};

/**
 * Helper type to differentiate a Mutation from a Query.
 *
 * @deprecated
 */
export type Mutation<P, R> = Operation<P, R> & {
  /**
   * Phantom type (never assigned) for indicating it is a mutation.
   */
  readonly _O: "mutation";
};

/**
 * A typed GraphQL operation with defined parameters and response.
 *
 * The `_*` properties never exist, they only carry type information,
 * also called phantom data. This type-information is used to enforce
 * operation-parameters and responses at compile-time.
 *
 * An Operation instance can only be constructed using type-assertions.
 *
 * Example:
 *
 *   interface MyParamerters {
 *     path: string;
 *   }
 *
 *   interface MyResponse {
 *     page: {
 *       contents: string;
 *     }
 *   }
 *
 *   const myQuery = someDocumentNode as Operation<MyParameters, MyResponse>;
 *
 *   runQuery(myQuery, { path: "foo" }).page.contents // OK!
 *   runQuery(myQuery, { path: ["foo"] }) // error
 *   runQuery(myQuery, { path: "foo" }).contents // error
 *
 * @see OperationParameters -- Helper to obtain the variables type
 * @see OperationResult -- Helper to obtain the result type
 *
 * @deprecated - Use TypedDocumentNode from @graphql-typed-document-node/core
 */
export type Operation<P, R> = TypedDocumentNode<R, P>;

/**
 * The parameters of an operation of type O.
 *
 * @deprecated - Use VariablesOf from @graphql-typed-document-node/core
 */
export type OperationParameters<O> = VariablesOf<O>;

/**
 * The results of an operation of type O.
 *
 * @deprecated - Use ResultOf from @graphql-typed-document-node/core
 */
export type OperationResult<O> = ResultOf<O>;

/**
 * Type describing an object which cannot have any properties ever.
 */
export type EmptyObject = Record<string | number | symbol, never>;

/**
 * Map of name in AST -> name in operation.
 *
 * @internal
 */
export type RenameMap = Readonly<Record<string, string>>;

/**
 * A basic GraphQL error received from the server.
 */
export interface GraphQLError {
  message: string;
  path?: string[];
}

/**
 * A basic GraphQL response.
 */
export type GraphQLResponse<T> =
  | {
      errors?: GraphQLError[];
      data: T;
    }
  | {
      errors: GraphQLError[];
    };
