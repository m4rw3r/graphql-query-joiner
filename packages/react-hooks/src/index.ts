import type {
  Client,
  GraphQLError,
  OptionalParameterIfEmpty,
  QueryError,
  ResultOf,
  TypedDocumentNode,
  VariablesOf,
} from "@awardit/graphql-ast-client";

import {
  createContext,
  useCallback,
  useContext,
  useMemo,
  useState,
} from "react";
import { createSharedState } from "@m4rw3r/react-pause-champ";

/**
 * Callback which executes a lazy operation.
 *
 * @public
 * @typeParam O - Typed GraphQL operation
 */
export interface ExecuteOperationCallback<
  O extends TypedDocumentNode<any, any>,
> {
  /**
   * Executes the operation with the given arguments, will suspend the
   * component during the request.
   */
  (...args: OptionalParameterIfEmpty<VariablesOf<O>>): void;
  /**
   * Resets the contents of the operation result, triggers re-render.
   */
  reset(): void;
}

/**
 * A result which can have partially failed, with erroring fields being
 * replaced by nulls.
 *
 * @public
 * @typeParam T - Query result type
 */
export interface FallibleResult<T> {
  /**
   * Query result, nullable fields can possibly be null due to query errors
   * present in the errors property.
   */
  data: T;
  /**
   * List of query errors which occurred during the query.
   */
  errors: GraphQLError[];
}

/**
 * Extra properties on a query result.
 *
 * @public
 */
interface UseQueryExtra {
  /**
   * Refetches the query, suspends the component during the request.
   */
  refetch: () => void;
}

/**
 * @internal
 */
type InnerLazyData<T> =
  | ["pending", Promise<unknown>]
  | ["data", T]
  | ["error", unknown];

/**
 * @internal
 */
const QUERY_PREFIX = "query:";
/**
 * @internal
 */
const FALLIBLE_QUERY_PREFIX = "query!:";

/**
 * @internal
 */
export const context = createContext<Client | undefined>(undefined);
/**
 * Provider for the GraphQL Client.
 *
 * @public
 */
export const Provider = context.Provider;

/**
 * Hook which returns the current graphql-client.
 *
 * @throws Error
 * This exception is thrown if no client was provided by a <Provider/> component.
 *
 * @public
 * @see {@link Provider}
 */
export function useClient(): Client {
  const client = useContext(context);

  if (!client) {
    throw new Error("useClient() must be used inside a <Provider/>");
  }

  return client;
}

/**
 * Hook which fetches a result of a GraphQL Query and suspends the rendering
 * during the request, errors will be thrown to the closest ErrorBoundary.
 *
 * NOTE: Operation-names should be unique for each query.
 *
 * Works with server-rendering using @m4rw3r/react-pause-champ.
 *
 * @public
 * @typeParam Q - Typed GraphQL query
 * @see {@link @m4rw3r/react-pause-champ:useChamp}
 * @see {@link @m4rw3r/react-pause-champ:Provider}
 * @see {@link @m4rw3r/react-pause-champ:Resume}
 * @see {@link https://react.dev/reference/react/Suspense}
 */
export function useQuery<Q extends TypedDocumentNode<any, any>>(
  query: Q,
  ...args: OptionalParameterIfEmpty<VariablesOf<Q>>
): ResultOf<Q> & UseQueryExtra {
  const client = useClient();
  // TODO: How to specify the name as a parameter?
  const id =
    QUERY_PREFIX + getQueryName(query) + ":" + JSON.stringify(args[0] ?? "");

  const [value, update] = createSharedState<ResultOf<Q>>(id)(() =>
    client(query, ...args),
  );
  const refetch = useCallback(() => {
    update(client(query, ...args));
  }, [update, query, args[0]]);

  // SAFETY: Value is always an object in a successful GraphQL response
  (value as UseQueryExtra).refetch = refetch;

  // SAFETY: By default we have any here from TypedDocumentNode<any, _>, but
  // should never happen in practice provided Q is fully typed
  //
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return value as ResultOf<Q> & UseQueryExtra;
}

/**
 * Hook which fetches the result, or partial result and query-errors, of a
 * GraphQL Query and suspends the rendering during the request.
 *
 * NOTE: Operation-names should be unique for each query.
 *
 * Works with server-rendering using @m4rw3r/react-pause-champ.
 *
 * @public
 * @typeParam Q - Typed GraphQL query
 * @see {@link useQuery}
 */
export function useFallibleQuery<Q extends TypedDocumentNode<any, any>>(
  query: Q,
  ...args: OptionalParameterIfEmpty<VariablesOf<Q>>
): FallibleResult<ResultOf<Q>> & UseQueryExtra {
  // TODO: Refactor

  const client = useClient();
  // TODO: How to specify the name as a parameter?
  const id =
    FALLIBLE_QUERY_PREFIX +
    getQueryName(query) +
    ":" +
    JSON.stringify(args[0] ?? "");

  const [value, update] = createSharedState<FallibleResult<ResultOf<Q>>>(id)(
    () => makeFallible(client(query, ...args)),
  );
  const refetch = useCallback(() => {
    update(makeFallible(client(query, ...args)));
  }, [update, query, args[0]]);

  (value as unknown as UseQueryExtra).refetch = refetch;

  return value as FallibleResult<ResultOf<Q>> & UseQueryExtra;
}

/**
 * Hook which manages an interactive query or mutation and its state.
 *
 * Initially the value will be undefined until the returned callback is called,
 * then the component will suspend during the operation until the operation has
 * completed.
 *
 * NOTE: Should never be called during the render of the component.
 *
 * @public
 * @typeParam O - Typed GraphQL operation
 */
export function useLazyOperation<O extends TypedDocumentNode<any, any>>(
  operation: O,
): [FallibleResult<ResultOf<O>> | undefined, ExecuteOperationCallback<O>] {
  const client = useClient();
  // We can store our promise in this state since the component does not throw
  // or trigger the operation during the initial render. This means that our
  // component state will be intact across a suspended promise.
  // TODO: Test with <React.StrictMode/>
  // TODO: Variant which throws on query errors as well?
  const [data, update] = useState<
    InnerLazyData<FallibleResult<ResultOf<O>>> | undefined
  >(undefined);
  const runOperation = useMemo<ExecuteOperationCallback<O>>(() => {
    const fn = (...args: OptionalParameterIfEmpty<VariablesOf<O>>) => {
      // Should only be updated on client
      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
      if (typeof process?.release?.name !== "undefined") {
        throw new Error(
          `Lazy GraphQL callbacks should only be called in a client environment`,
        );
      }

      // TODO: Replace by an Entry?
      const promise = makeFallible(client(operation, ...args)).then(
        (result) => {
          update(["data", result]);
        },
        (error) => {
          update(["error", error]);
        },
      );

      // just suspend
      update(["pending", promise]);
    };

    fn.reset = function reset(): void {
      update(undefined);
    };

    return fn;
  }, [client, operation, update]);

  if (data && data[0] !== "data") {
    // Not data, so either promise or error
    throw data[1];
  }

  return [data?.[1], runOperation];
}

/**
 * @internal
 */
function makeFallible<T>(promise: Promise<T>): Promise<FallibleResult<T>> {
  return promise.then(
    (data) => ({ data, errors: [] }),
    (error) => {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      if (typeof error === "object" && error?.name === "QueryError") {
        // SAFETY: We trust that the Query error data actually is our data
        return {
          data: (error as QueryError).queryData as T,
          errors: (error as QueryError).errors,
        };
      } else {
        throw error;
      }
    },
  );
}

/**
 * @internal
 */
function getQueryName(query: TypedDocumentNode<any, any>): string {
  if (!query.definitions[0] || !("name" in query.definitions[0])) {
    throw new Error("Query is missing name");
  }

  return query.definitions[0].name.value;
}
