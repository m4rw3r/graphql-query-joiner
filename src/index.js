/* @flow */

export type {
  Query,
  QueryParameters,
  QueryResponse,
  GraphQLError,
  GraphQLResponse,
} from "./query";
export type { Client, ClientArgs, QueryRunner } from "./client";

export { createClient, handleResponse } from "./client";

