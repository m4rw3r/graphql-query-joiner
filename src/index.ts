export type {
  Query,
  QueryParameters,
  QueryResponse,
  GraphQLError,
  GraphQLResponse,
} from "./query";
export type { Client, ClientArgs, QueryRunner } from "./client";
export { createClient, handleFetchResponse } from "./client";
