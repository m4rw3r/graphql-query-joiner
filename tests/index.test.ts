import { graphql } from "../src";
import { rollup } from "rollup";

process.chdir(__dirname);

test("simple", async () => {
  const bundle = await rollup({
    input: "fixtures/simple/main.js",
    plugins: [
      graphql(),
    ],
  });

  const { output: [{ code }] } = await bundle.generate({
    format: "esm"
  });

  expect(code).toMatchSnapshot();
});

test("fragment", async () => {
  const bundle = await rollup({
    input: "fixtures/fragment/main.js",
    plugins: [
      graphql(),
    ],
  });
  const { output: [{ code }] } = await bundle.generate({
    format: "esm"
  });

  expect(code).toMatchSnapshot();
});

test("schema", async t => {
  const bundle = await rollup({
    input: "fixtures/schema/main.js",
    plugins: [
      graphql(),
    ],
  });
  const { output: [{ code }] } = await bundle.generate({
    format: "esm"
  });

  expect(code).toMatchSnapshot();
});