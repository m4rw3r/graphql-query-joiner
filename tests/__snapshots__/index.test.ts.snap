// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fragment 1`] = `
"const User = {
  kind: "Document",
  definitions: [{"kind":"FragmentDefinition","name":{"kind":"Name","value":"User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"},"arguments":[],"directives":[]}]}}]
};
const foo = {
  kind: "Document",
  definitions: [].concat(User.definitions, {"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"foo"},"variableDefinitions":[],"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"users"},"arguments":[],"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"FragmentSpread","name":{"kind":"Name","value":"User"},"directives":[]}]}}]}})
};

console.log(foo);
console.log(User);
"
`;

exports[`schema 1`] = `
"const User = {
  kind: "Document",
  definitions: [{"kind":"ObjectTypeDefinition","name":{"kind":"Name","value":"User"},"interfaces":[],"directives":[],"fields":[{"kind":"FieldDefinition","name":{"kind":"Name","value":"name"},"arguments":[],"type":{"kind":"NonNullType","type":{"kind":"NamedType","name":{"kind":"Name","value":"String"}}},"directives":[]}]}]
};
const typeDefs = {
  kind: "Document",
  definitions: [{"kind":"ObjectTypeDefinition","name":{"kind":"Name","value":"User"},"interfaces":[],"directives":[],"fields":[{"kind":"FieldDefinition","name":{"kind":"Name","value":"name"},"arguments":[],"type":{"kind":"NonNullType","type":{"kind":"NamedType","name":{"kind":"Name","value":"String"}}},"directives":[]}]}, {"kind":"ObjectTypeDefinition","name":{"kind":"Name","value":"Query"},"interfaces":[],"directives":[],"fields":[{"kind":"FieldDefinition","name":{"kind":"Name","value":"user"},"arguments":[],"type":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"directives":[]}]}, {"kind":"ObjectTypeDefinition","name":{"kind":"Name","value":"Mutation"},"interfaces":[],"directives":[],"fields":[{"kind":"FieldDefinition","name":{"kind":"Name","value":"updateUser"},"arguments":[],"type":{"kind":"NonNullType","type":{"kind":"NamedType","name":{"kind":"Name","value":"Boolean"}}},"directives":[]}]}]
};

console.log(typeDefs);
console.log(User);
"
`;

exports[`simple 1`] = `
"const foo = {
  kind: "Document",
  definitions: [{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"foo"},"variableDefinitions":[],"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"info"},"arguments":[],"directives":[]}]}}]
};
const bar = {
  kind: "Document",
  definitions: [{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"bar"},"variableDefinitions":[],"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"more"},"arguments":[],"directives":[]}]}}]
};

console.log(foo);
console.log(bar);
"
`;
